lua/lsp/null-ls.lua

```diff
@@ -1,141 +1,136 @@
-local M = {}
-local Log = require "core.log"
-
+local M = {
+  langs = {
+    -- [...] = {
+    --   registered = {
+    --     formatters = {},
+    --     linters = {},
+    --   },
+    --   errors = {
+    --     formatters = {},
+    --     linters = {},
+    --   },
+    -- },
+  },
+}
+
+local logger = require("core.log"):get_default()
 local null_ls = require "null-ls"
 
-local nodejs_local_providers = { "prettier", "prettierd", "prettier_d_slim", "eslint_d", "eslint" }
+local function find_local_node_package(root_dir, command)
+  return root_dir .. "/node_modules/.bin/" .. command
+end
 
-M.requested_providers = {}
+local local_installations = {
+  prettier = find_local_node_package,
+  prettierd = find_local_node_package,
+  prettier_d_slim = find_local_node_package,
+  eslint_d = find_local_node_package,
+  eslint = find_local_node_package,
+}
 
-function M.get_registered_providers_by_filetype(ft)
-  local matches = {}
-  for _, provider in pairs(M.requested_providers) do
-    if vim.tbl_contains(provider.filetypes, ft) then
-      local provider_name = provider.name
-      -- special case: show "eslint_d" instead of eslint
-      -- https://github.com/jose-elias-alvarez/null-ls.nvim/blob/9b8458bd1648e84169a7e8638091ba15c2f20fc0/doc/BUILTINS.md#eslint
-      if string.find(provider._opts.command, "eslint_d") then
-        provider_name = "eslint_d"
-      end
-      table.insert(matches, provider_name)
-    end
+local function find_root_dir()
+  if lvim.builtin.rooter.active then
+    --- use vim-rooter to set root_dir
+    vim.cmd "let root_dir = FindRootDirectory()"
+    return vim.api.nvim_get_var "root_dir"
   end
 
-  return matches
-end
+  -- TODO: Rework this to not make it javascript specific
+  --- use LSP to set root_dir
+  local ts_client = require("utils").get_active_client_by_ft "typescript"
+  if ts_client == nil then
+    logger.error "Unable to determine root directory since tsserver didn't start correctly"
+    return nil
+  end
 
-function M.get_missing_providers_by_filetype(ft)
-  local matches = {}
-  for _, provider in pairs(M.requested_providers) do
-    if vim.tbl_contains(provider.filetypes, ft) then
-      local provider_name = provider.name
+  return ts_client.config.root_dir
+end
 
-      table.insert(matches, provider_name)
+local function find_command(command)
+  if local_installations[command] then
+    local root_dir = find_root_dir()
+    local local_command = local_installations[command](root_dir, command)
```

Comment:
I feel like this is abstraction in the wrong direction. 
It's kinda confusing that `local_installations` is a list of callbacks instead of the actual path. From a glance you wouldn't guess that it would be. 
Secondly, why is it called `local_installation` when in fact it's relative to the `root_dir`?  where one  would expect it to be any arbitrary path that's set somehow. Instead there's the hidden dependency that the first arg needs to actually be fed as start point so that it can find it.

I'm not sure how this is better than using the clearly defined `validate_nodejs_providers`. Another side effect, is that we would lose the ability to log those failed register attempts more granularly.